packages/@lwc/engine-core/src/renderer.ts

```diff
@@ -10,267 +13,75 @@ export type HostElement = any;
-export function setGetChildNodes(getChildNodesImpl: getChildNodesFunc) {
-    getChildNodes = getChildNodesImpl;
-}
-
-type getFirstChildFunc = (element: E) => N | null;
-export let getFirstChild: getFirstChildFunc;
-export function setGetFirstChild(getFirstChildImpl: getFirstChildFunc) {
-    getFirstChild = getFirstChildImpl;
-}
-
-type getFirstElementChildFunc = (element: E) => E | null;
-export let getFirstElementChild: getFirstElementChildFunc;
-export function setGetFirstElementChild(getFirstElementChildImpl: getFirstElementChildFunc) {
-    getFirstElementChild = getFirstElementChildImpl;
-}
-
-type getLastChildFunc = (element: E) => N | null;
-export let getLastChild: getLastChildFunc;
-export function setGetLastChild(getLastChildImpl: getLastChildFunc) {
-    getLastChild = getLastChildImpl;
-}
-
-type getLastElementChildFunc = (element: E) => E | null;
-export let getLastElementChild: getLastElementChildFunc;
-export function setGetLastElementChild(getLastElementChildImpl: getLastElementChildFunc) {
-    getLastElementChild = getLastElementChildImpl;
-}
-
-type isConnectedFunc = (node: N) => boolean;
-export let isConnected: isConnectedFunc;
-export function setIsConnected(isConnectedImpl: isConnectedFunc) {
-    isConnected = isConnectedImpl;
-}
-type insertStylesheetFunc = (content: string, target?: ShadowRoot) => void;
-export let insertStylesheet: insertStylesheetFunc;
-export function setInsertStylesheet(insertStylesheetImpl: insertStylesheetFunc) {
-    insertStylesheet = insertStylesheetImpl;
-}
-
-type assertInstanceOfHTMLElementFunc = (elm: any, msg: string) => void;
-export let assertInstanceOfHTMLElement: assertInstanceOfHTMLElementFunc;
-export function setAssertInstanceOfHTMLElement(
-    assertInstanceOfHTMLElementImpl: assertInstanceOfHTMLElementFunc
-) {
-    assertInstanceOfHTMLElement = assertInstanceOfHTMLElementImpl;
-}
-
-type defineCustomElementFunc = (
-    name: string,
-    constructor: CustomElementConstructor,
-    options?: ElementDefinitionOptions
-) => void;
-export let defineCustomElement: defineCustomElementFunc;
-export function setDefineCustomElement(defineCustomElementImpl: defineCustomElementFunc) {
-    defineCustomElement = defineCustomElementImpl;
-}
-
-type getCustomElementFunc = (name: string) => CustomElementConstructor | undefined;
-export let getCustomElement: getCustomElementFunc;
-export function setGetCustomElement(getCustomElementImpl: getCustomElementFunc) {
-    getCustomElement = getCustomElementImpl;
+let defaultRenderer: RendererAPI;
+
+export function setDefaultRenderer(renderer: RendererAPI) {
+    if (defaultRenderer !== undefined) {
+        throw Error(`Invalid Renderer Initialization`);
+    }
+    defaultRenderer = renderer;
+}
+
+export function getRendererFromVNode(vnode: VBaseElement): RendererAPI {
+    return vnode.data.renderer || defaultRenderer;
+}
+
+export function getRendererFromVM(vm: VM): RendererAPI {
+    return vm.renderer || defaultRenderer;
```

Comment:
The `VM.renderer` is never nullish. No need for the extra lookup to the default renderer.

```suggestion
    return vm.renderer;
```